include/xtensor/xmasked_value.hpp

```diff
@@ -0,0 +1,817 @@
+                m_value OP value;                                                             \
+            }                                                                                 \
+            return *this;                                                                     \
+        }                                                                                     \
+                                                                                              \
+        template <class T1, class B1>                                                         \
+        inline xmasked_value& operator OP(const xtl::xoptional<T1, B1>& value)                \
+        {                                                                                     \
+            m_flag = m_flag && value.has_value();                                             \
+            if (m_flag)                                                                       \
+            {                                                                                 \
+                m_value OP value.value();                                                     \
+            }                                                                                 \
+            return *this;                                                                     \
+        }                                                                                     \
+                                                                                              \
+        template <class T1, class B1>                                                         \
+        inline xmasked_value& operator OP(const xmasked_value<T1, B1>& value)                 \
+        {                                                                                     \
+            m_flag = m_flag && value.has_value();                                             \
+            if (m_flag)                                                                       \
+            {                                                                                 \
+                m_value OP value.value();                                                     \
+            }                                                                                 \
+            return *this;                                                                     \
+        }
+
+        DEFINE_ASSIGN_OPERATOR(=);
+        DEFINE_ASSIGN_OPERATOR(+=);
+        DEFINE_ASSIGN_OPERATOR(-=);
+        DEFINE_ASSIGN_OPERATOR(*=);
+        DEFINE_ASSIGN_OPERATOR(/=);
+        DEFINE_ASSIGN_OPERATOR(%=);
+        DEFINE_ASSIGN_OPERATOR(&=);
+        DEFINE_ASSIGN_OPERATOR(|=);
+        DEFINE_ASSIGN_OPERATOR(^=);
+
+#undef DEFINE_ASSIGN_OPERATOR
+
+        inline void swap(xmasked_value& other)
+        {
+            std::swap(m_value, other.m_value);
+            std::swap(m_flag, other.m_flag);
```

Comment:
This prevents using custom type as the `value_type` (and potentially the `flag_type`) for which a `swap` method has been defined.. Using a not full qualified names gives more flexibility:

```cpp
inline void swap(xmasked_value& other)
{
    using std::swap;
    swap(m_value, other.m_value);
    swap(m_flag, other.m_flag);
}
```

Also do not forget the free function binding on the method (for ADL resolution).